app/validators/questionnaire_validator.py

```diff
@@ -0,0 +1,363 @@
+
+        for required_section_id in required_section_ids:
+            if required_section_id not in section_ids:
+                self.add_error(
+                    error_messages.REQUIRED_HUB_SECTION_UNDEFINED,
+                    required_section_id=required_section_id,
+                )
+
+    def validate_blocks(self, section_id, group_id, numeric_answer_ranges):
+        section = self.questionnaire_schema.get_section(section_id)
+        group = self.questionnaire_schema.get_group(group_id)
+
+        for block in group.get("blocks"):
+            if (
+                section_id == self.schema_element["sections"][-1]["id"]
+                and group_id == section["groups"][-1]["id"]
+                and block["id"] == group["blocks"][-1]["id"]
+            ):
+                self.validate_block_is_submission(block)
+
+            block_routing_validator = RoutingValidator(
+                block, group, self.questionnaire_schema
+            )
+            block_routing_validator.validate()
+            self.errors += block_routing_validator.errors
+
+            block_validator = get_block_validator(block, self.questionnaire_schema)
+            block_validator.validate()
+            self.errors += block_validator.errors
+
+            self.validate_questions(block, numeric_answer_ranges)
+            self.validate_variants(block, numeric_answer_ranges)
+
+    def validate_questions(self, block_or_variant, numeric_answer_ranges):
+        questions = block_or_variant.get("questions", [])
+        question = block_or_variant.get("question")
+        routing_rules = block_or_variant.get("routing_rules", {})
+
+        if question:
+            questions.append(question)
+
+        for question in questions:
+            question_validator = get_question_validator(question)
+            question_validator.validate()
+
+            self.errors += question_validator.errors
+
+            for answer in question.get("answers", []):
+                if routing_rules:
+                    answer_routing_validator = AnswerRoutingValidator(
+                        answer, routing_rules
+                    )
+                    answer_routing_validator.validate()
+                    self.errors += answer_routing_validator.errors
+
+                answer_validator = get_answer_validator(
+                    answer,
+                    self.questionnaire_schema.list_names,
+                    self.questionnaire_schema.block_ids,
+                )
+
+                answer_validator.validate()
+
+                if isinstance(answer_validator, NumberAnswerValidator):
+                    numeric_answer_ranges[
+                        answer["id"]
+                    ] = answer_validator.get_numeric_range_values(numeric_answer_ranges)
+
+                    answer_validator.validate_numeric_answer_types(
+                        numeric_answer_ranges
+                    )
```

Comment:
numeric_answer_ranges is built from answers across the entire schema - the validation depends on the ranges of previous questions having been returned.

Reply:
This seems like it would be better suited to work out the answer ranges up front, and make them available to the answer validator.

Reply:
Yes, I'd considered this too. It results in iterating everything twice and repeating the validator creation. That approach seems worse than this.

Reply:
I don't think iterating twice is an issue. One of the big problems with the original validator is complexity because it tries to be efficient rather than clearly understood. The validator wouldn't be created until all of the information it requires is gathered, so I can't see why the validator would need to be created twice?

Reply:
I've refactored this so the getter now lives in the QuestionnaireSchema and is called ahead of time